flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/plan/rules/logical/PushProjectIntoTableSourceScanRule.java

```diff
@@ -300,7 +300,15 @@ private RowType performPushDown(
                                 .getLogicalType();
 
         if (supportsProjectionPushDown(source.tableSource())) {
-            abilitySpecs.add(new ProjectPushDownSpec(physicalProjections, newProducedType));
+            abilitySpecs.add(
+                    new ProjectPushDownSpec(
+                            physicalProjections,
+                            (RowType)
+                                    DataTypeUtils.projectRow(
```

Comment:
move `newProducedType` after this `if` branch and introduce a local variable with a helpful name e.g. `RowType projectedPhysicalType`